File: AutoPlaque-QSM-python/utils/extract_slice.py
"""
extract_slice - extract a 2-D sagittal frame from 3-D MRI volumes.
"""

import numpy as np


def extract_slice(volume_img: np.ndarray, volume_lbl: np.ndarray, slice_idx: int):
    """
    Extract a 2-D sagittal frame and its corresponding label mask.

    Parameters
    ----------
    volume_img : 3-D ndarray (float/double), MRI volume
    volume_lbl : 3-D ndarray (int/uint16), label volume
    slice_idx  : int, 1-based index along the second (y) axis (MATLAB convention)

    Returns
    -------
    slice_frame : 2-D ndarray, dtype float64
    slice_mask  : 2-D ndarray, dtype uint16
    """
    n_rows, n_cols, n_slices = volume_img.shape
    # Convert 1-based MATLAB index to 0-based Python index
    j = slice_idx - 1

    if j < 0 or j >= n_cols:
        raise IndexError(
            f"slice_idx {slice_idx} is out of bounds for volume with {n_cols} columns."
        )

    slice_frame = volume_img[:, j, :].astype(np.float64)
    slice_mask = volume_lbl[:, j, :].astype(np.uint16)

    if slice_frame.size == 0:
        import warnings
        warnings.warn(f"Extracted frame at index {slice_idx} is empty.")

    return slice_frame, slice_mask

File: utils/extractSlice.m
function [sliceFrame, sliceMask] = extractSlice(volumeImg, volumeLbl, sliceIdx)
% EXTRACTSLICE Extracts a 2D sagittal frame and its corresponding label mask.
%
% This function performs dimension reduction from 3D to 2D while ensuring 
% numerical precision for QSM analysis and categorical integrity for labels.
%
% INPUTS:
%   volumeImg - 3D double/float matrix (MRI Volume)
%   volumeLbl - 3D uint16/categorical matrix (Label Volume)
%   sliceIdx  - Integer, the index of the column (y-axis) to extract
%
% OUTPUTS:
%   sliceFrame - 2D double matrix (Ready for GoDec/RXD)
%   sliceMask  - 2D uint16 matrix (ROI Mask)

    %% 1. Dimensionality and Index Validation
    [nRows, nCols, nSlices] = size(volumeImg);

    if sliceIdx > nCols || sliceIdx < 1
        error('QSM:ExtractSlice:BoundsError', ...
              'Index %d is out of bounds for volume with %d columns.', sliceIdx, nCols);
    end

    %% 2. Extraction and Orientation Management
    % Using squeeze to remove the singleton dimension. 
    % Note: In your pipeline, slicing is performed along the 2nd dimension (y-axis).
    try
        sliceFrame = double(squeeze(volumeImg(:, sliceIdx, :)));
        sliceMask  = uint16(squeeze(volumeLbl(:, sliceIdx, :)));
    catch ME
        error('QSM:ExtractSlice:RuntimeError', ...
              'Failed to extract slice at index %d: %s', sliceIdx, ME.message);
    end

    %% 3. Post-Extraction Integrity Check
    % Ensure the frame is not empty and dimensions are consistent
    if isempty(sliceFrame)
        warning('QSM:ExtractSlice:EmptyFrame', 'Extracted frame at index %d is empty.', sliceIdx);
    end
end